### PR TITLE
Switch CI checks to github actions from travis

### PR DESCRIPTION
**Pull request by rwx788**
Unfortunately, we face enormous delays when jobs are picked up by travis CI. It's wrong to complain about it, as these services are provided for free. Nevertheless, we had an idea of switching to github actions due to other advantages and better integration with github services, like using organization secrets instead of defining them in each repo.

For the public repos github actions are completely free.

For podman documentation, there was a change done, that is generated unconditionally on merge, as deploy plugin will psh commit to gh-pages only if there was a change. This step also will be done only on PR merge and not triggered on push to the forked branch.

No changes done to the compilation check, so it still conducted for all the perl modules in the repo and takes a lot of time. This will be addresses in the separate PR, so we run compile check for the changed files only and full check once PR is merged or just once a day, which should be enough to keep repo in a consistent state.

#### Test runs:
* [Run with broken yaml](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11545/checks?check_run_id=1498325977).

### CODE REVIEW

File: .github/workflows/static.yml
Comment (okurz) [inline .github/workflows/static.yml:34]: I consider it easier to test and also easier to apply for different CIs when you can extract the script sections into individual scripts. Compare to https://github.com/os-autoinst/os-autoinst/blob/master/.travis.yml#L6 , only 6 lines in that file :)

File: .github/workflows/static.yml
Comment (okurz) [inline .github/workflows/static.yml:15]: not sure if you extend this step by step and I assume you are aware but just to ensure nothing goes missed: So far we have three parallel jobs "static", "unit", "compile", here only two are specified so far.

File: .github/workflows/static.yml
Comment (rwx788) [reply .github/workflows/static.yml]: Let me make it works as it was in travis before we start with improvements ;)

File: .github/workflows/static.yml
Comment (rwx788) [reply .github/workflows/static.yml:15]: Yes, one of the things I would do straight away is to do compile check on changed files only, and full check once merged or even just once a day.
Also generation of perl doc is not added yet as will need more changes to make it work.

File: .github/workflows/static.yml
Comment (rwx788) [reply .github/workflows/static.yml:34]: Also, do not forget that we need some budget for it too, as 2000 minutes per 20 days doesn't seem to be enough for us, even if we improve CI to run faster (now due to compile check for all files it takes 10+ minutes).
So I just want to make first step in case we decide to switch.

Comment (okurz) [reply .github/workflows/static.yml]: yeah sure. What I actually meant is to first *refactor* what we have in travis CI, then have simple files which can be called in parallel in both github actions and travis CI so that we can also test and see which one is faster (or which one exceeds free quota first) and then drop one

File: .github/workflows/static.yml
Comment (okurz) [reply .github/workflows/static.yml:15]: Let me quote yourself from https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11545#discussion_r535345488

"Let me make it works as it was in travis before we start with improvements ;)"

File: Makefile
Comment (b10n1k) [inline Makefile:63]: maybe you could also change the `y.\?ml` as practically it will match also yeml,y222ml etc
Also from the man page
```
git diff [<options>] <commit>...<commit> [--] [<path>...]
           This form is to view the changes on the branch containing and up to the second <commit>,
           starting at a common ancestor of both <commit>. "git diff A...B" is equivalent to "git
           diff $(git merge-base A B) B". You can omit any one of <commit>, which has the same effect
           as using HEAD instead.
``` seems simplier

Comment (rwx788) [reply Makefile:63]: I will focus on making it work first, I cannot improve all the problems we have in Makefile in 5 minutes. Feel free to create separate PR with this change.

Comment (rwx788) [reply Makefile:63]: As for diff option, I've tested it and indeed we can put branch name, so will simplify that part.

File: .github/workflows/static.yml
Comment (rwx788) [reply .github/workflows/static.yml]: Please, check PR description, github actions are for free for public repos. And main issue is not that it takes long to run (even though we can improve there), but that jobs now are picked up very late, e.g. here last push was 3 hours ago: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11551
And I had cases when after 10 hours CI still would not run. That's why I decided to act on this now as it really affects our daily work.

Comment (okurz) [reply .github/workflows/static.yml]: Of course. No doubt about that. I appreciate that you prepare a switch to github actions. By the way, I am also doing this: https://github.com/os-autoinst/os-autoinst/pull/1576 :)

**Comment by foursixnine** [comment]
Looks really good so far @rwx788!

**Comment by rwx788** [comment]
> Looks really good so far @rwx788!

So I would wait till this PR is merged before proceeding with further changes. Same functionality as we had before works. I didn't test documentation deploy properly as GITHUB_TOKEN is not available by default for the CI runs from the forks, but I'm quite confident about the configuration.

**Review by okurz — APPROVED** [review]
I am not sure about all the changes you did but it looks a bit cleaner than before even :)

I am wondering about the change from "FETCH_HEAD" to "origin/master" though this might be needed when github actions create the temporary state for tests different than travis CI.

Overall looks very good and thank you.

**Comment by rwx788** [comment]
> I am not sure about all the changes you did but it looks a bit cleaner than before even :)
> 
> I am wondering about the change from "FETCH_HEAD" to "origin/master" though this might be needed when github actions create the temporary state for tests different than travis CI.
I was also not sure about it, on this picture it's cleat though: 
image

Since version 1.8.4 git fetch origin master won't only update FETCH_HEAD, but also origin/master, so we can use it without a fear that `FETCH_HEAD` changes it's state and I personally think it's more transparent.

> 
> Overall looks very good and thank you.

**Comment by rwx788** [comment]
> I am not sure about all the changes you did but it looks a bit cleaner than before even :)
> 
> I am wondering about the change from "FETCH_HEAD" to "origin/master" though this might be needed when github actions create the temporary state for tests different than travis CI.

> 
> Overall looks very good and thank you.
Cheers!

**Comment by okurz** [comment]
ah, cool. Something learned :)

**Review by OleksandrOrlov — APPROVED** [review]
Great!

**Comment by foursixnine** [comment]
This is about removing travis :D

File: .github/workflows/deploy.yml
Comment (Vogtinator) [inline .github/workflows/deploy.yml:16]: For security reasons, it's highly recommended to pin this to a commit hash instead of a tag. Otherwise you give a third-party full access to the repository!

**Comment by okurz** [comment]
Other tasks I suggest as followup: remove all references to "travis CI", e.g. the badge on README

File: .github/workflows/deploy.yml
Comment (rwx788) [reply .github/workflows/deploy.yml:16]: We need a separate account with limited permissions here. Problem with travis looked much more urgent and we had exact same issue there. Will add it to follow-up tasks. To calm everyone here, those plugins are open source and reviewed for any malicious code before being put to the market place, and if you don't trust third party components in 21-st century, you should implement your own github =)

Comment (Vogtinator) [reply .github/workflows/deploy.yml:16]: > To calm everyone here, those plugins are open source and reviewed for any malicious code before being put to the market place

That's not true - you can use any github repo as action and anyone with write access to tags can just insert malicious code without any review.

https://julienrenaux.fr/2019/12/20/github-actions-security-risk/

The fix is simple:

```suggestion
      uses: peaceiris/actions-gh-pages@bbdfb200618d235585ad98e965f4aafc39b4c501
```

> and if you don't trust third party components in 21-st century, you should implement your own github =)

Well, this is equivalent to just giving anyone with push access to the action maintainer access to this repo...

Comment (rwx788) [reply .github/workflows/deploy.yml:16]: I do not argue with your point, I've stated that we should ideally have token which cannot do any malicious actions. There are still a lot of `if` in the described case. We also need commit for the `checkout` action. If you could provide PR with those changes, would be great. If not, I will take care of it later today/tomorrow.

Comment (Vogtinator) [reply .github/workflows/deploy.yml:16]: The `checkout` action is actually reviewed by GitHub, so your argument about trust applies there.

I'll make a PR, but it seems like @foursixnine just put a syntax error into deploy.yml (`You have an error in your yaml syntax on line 13`)

Comment (foursixnine) [reply .github/workflows/deploy.yml:16]: We should be all good now, @Vogtinator gogogo!

**Pull request merged**
